#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
data_path = '../M3/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [2]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
df.groupby('Branch')[['Total','gross income']].sum().sort_values(by='Total', ascending=False)

Total  gross income
Branch                           
C       110568.7065     5265.1765
A       106200.3705     5057.1605
B       106197.6720     5057.0320

Відповідь: Найбільший дохід має філіал C. 
Для повноти аналізу виведено як загальна сума доходу від філіалів, так і валовий дохід.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
# Варіант 1 (коли ми знаємо, що назві філіалу відповідає одне унікальне місто)

df[df.Branch == 'C'].City.unique()

array(['Naypyitaw'], dtype=object)

In [5]:
# Варіант 2 (коли ми хочемо переконатися, що філіал з назвою C розташований в одному місті, а не наприклад в кількох)

df[df.Branch == 'C'].groupby('City')[['Total','gross income']].sum()

Total  gross income
City                                
Naypyitaw  110568.7065     5265.1765

Відповідь: Філіал з найвищим прибутком знаходиться в місті Naypyitaw

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [6]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [7]:
invoices_by_category.sort_values(by='Electronic accessories', ascending=False).head(1)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Гіпотеза 1. Аналіз профілю основного клієнта у розрізі міст по ключовим характеристикам:
- тип споживача (Customer type)
- стать (Gender)
- середня дохідність (Total)

In [8]:
pd.pivot_table(df, index=['Customer type','Gender'], columns='City', values=['Invoice ID','Total'], aggfunc={
    'Invoice ID': 'count',
    'Total': 'mean'})

Invoice ID                        Total              \
City                   Mandalay Naypyitaw Yangon    Mandalay   Naypyitaw   
Customer type Gender                                                       
Member        Female         85        96     80  315.887435  360.973375   
              Male           80        73     87  335.678175  304.490938   
Normal        Female         77        82     81  338.673545  329.658768   
              Male           90        77     92  293.501367  346.174091   

                                  
City                      Yangon  
Customer type Gender              
Member        Female  333.038344  
              Male    310.280552  
Normal        Female  328.717278  
              Male    281.921690

Відповідь: з попереднього аналізу знаємо, що найбільший дохід демонструє місто Naypyitaw (філія С). 
З даного аналізу видно, що такий дохід місто Naypyitaw отримує за рахунок аудиторії "Жінки з картками члена", що приносять точці максимальний середній дохід ~361 грн/чек. При цьому з таблиці видно, що інші міста Mandalay та Yangon мають найбільше клієнтів профілем "Чоловіки без картки", що приносять найменший по профілю середній дохід 293,5 грн. та 282 грн. відповідно.

Отже, рекомендація для міст Mandalay та Yangon - це фокус на залучення аудиторії Жінок та стимулювання оформлення для них карток членів.

Гіпотеза 2. Перевірити рівень цін на категорії продуктів у розрізі міст (чи відрізняється цінова політика міст)

In [9]:
pd.pivot_table(df, index='Product line', columns='City', values=['Total','Unit price'], aggfunc={
    'Total': 'sum',
    'Unit price': 'mean'})

Total                         Unit price  \
City                      Mandalay   Naypyitaw      Yangon   Mandalay   
Product line                                                            
Electronic accessories  17051.4435  18968.9745  18317.1135  49.854182   
Fashion accessories     16413.3165  21560.0700  16332.5085  54.843871   
Food and beverages      15214.8885  23766.8550  17163.1005  55.540000   
Health and beauty       19980.6600  16615.3260  12597.7530  58.185660   
Home and lifestyle      17549.1645  13895.5530  22417.1955  55.514000   
Sports and travel       19988.1990  15761.9280  19372.6995  59.678065   

                                              
City                    Naypyitaw     Yangon  
Product line                                  
Electronic accessories  55.809455  54.871167  
Fashion accessories     59.736000  56.670392  
Food and beverages      57.273030  54.974483  
Health and beauty       55.971346  49.862340  
Home and lifestyle      54.334222  55.845692  
Sports and travel       55.107333  55.610339

Відповідь: з даної таблиці ми бачимо, що найбільший дохід місту Naypyitaw приносять такі ТОП-3 категорії товарів: Food and beverages, Fashion accessories, Electronic accessories. При цьому саме ці категорії товарів мають найвищу ціну Unit price у місті Naypyitaw, порівняно з іншими містами Mandalay та Yangon. 

Отже, рекомендація для міст Mandalay та Yangon - переглянути цінову політику у своїх містах на товари категорій Food and beverages, Fashion accessories, Electronic accessories. Можливо вони потребують підвищення цін чи введення акційних цін лише на оптові замовлення та відміну акційних знижок на роздрібні покупки (якщо є така практика).

Гіпотеза 3. Перевірити тип розрахунків споживачів у розрізі міст по характеристикам:
- кількість споживачів
- їх середня дохідність

In [10]:
pd.pivot_table(df, index='Payment', columns='City', values=['Invoice ID','Total'], aggfunc={
    'Invoice ID': 'count',
    'Total': 'mean'})

Invoice ID                        Total                        
City          Mandalay Naypyitaw Yangon    Mandalay   Naypyitaw      Yangon
Payment                                                                    
Cash               110       124    110  321.267832  347.466593  307.102282
Credit card        109        98    104  342.613362  309.463929  318.218755
Ewallet            113       106    126  296.578354  350.522491  312.098167

Відповідь: аналіз показує, що місто Naypyitaw має найбільшу кількість споживачів, що розраховуються готівкою (Cash), при цьому середня дохідність їх досить висока 347,5 грошових одиниць і на рівні споживачів, що розраховуються за допомогою Ewallet. При цьому інші міста мають значно нижчу кількість таких споживачів - місто Mandalay має майже рівномірний спліт всіх 3-х типів споживачів, а місто Yangon має більшу кількість споживачів з розрахунком Ewallet.

Отже, рекомендація для міст Mandalay та Yangon - це фокус на залучення споживачів, що розраховуються готівкою, що можливо допоможе у підвищенні об'ємів продажів та доходів магазинів. 

Гіпотеза 4. Перевірити рівень задоволеності споживачів у розрізі міст

In [11]:
def rating_mapping(val):
    if val>=8:
        return 'Big'
    elif 6<=val<8:
        return 'Medium'
    else:
        return 'Small'

In [12]:
df['Rating Category'] = df.Rating.apply(rating_mapping)
df.groupby('City')['Rating Category'].value_counts(normalize=True)

City       Rating Category
Mandalay   Medium             0.367470
           Small              0.337349
           Big                0.295181
Naypyitaw  Big                0.362805
           Medium             0.341463
           Small              0.295732
Yangon     Medium             0.358824
           Big                0.329412
           Small              0.311765
Name: Rating Category, dtype: float64

Відповідь: в домашньому завданні 3.3 ми визначили шкалу задоволеності споживачів. Зараз ми розглянули розподіл цих сегментів (Rating Category) у розрізі міст. Аналіз показав, що лише у місті Naypyitaw переважають споживачі з Високим рівнем задоволення і становлять ~36,3%. Натомість у міст Mandalay та Yangon переважають споживачі з Середнім рівнем задоволення - 36,7% та 35,9% відповідно. При чому у місті Mandalay найгірші показники задоволеності - доля споживачів з Високим рівнем задоволення тут складає найменше 29,5%.

Отже, рекомендація для міст Mandalay та Yangon - переглянути рівень сервісу споживачів з метою покращення рівня задоволеності споживачів. Можливо потрібно провести опитування, анкетування клієнтів з метою визначення слабких зон магазину і їх виправлення.